incubator/hnc/pkg/reconcilers/hierarchical_namespace.go

```diff
@@ -31,44 +39,176 @@ type HierarchicalNamespaceReconciler struct {
 func (r *HierarchicalNamespaceReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	// TODO report error state if the webhook is bypassed - see issue
-	//  https://github.com/kubernetes-sigs/multi-tenancy/issues/459
-	if ex[req.Namespace] {
-		return ctrl.Result{}, nil
-	}
-
+	ctx := context.Background()
 	log := r.Log.WithValues("trigger", req.NamespacedName)
 	log.Info("Reconciling HNS")
 
-	// Names of the hierarchical namespace and the current namespace.
+	// Get names of the hierarchical namespace and the current namespace.
 	nm := req.Name
 	pnm := req.Namespace
 
-	// Set RequiredChildOf (Owner) in the forest of the hierarchical namespace to the
-	// current namespace.
-	// TODO rename RequiredChildOf to Owner in the forest. See issue:
-	//  https://github.com/kubernetes-sigs/multi-tenancy/issues/469
-	r.forest.Get(nm).RequiredChildOf = pnm
+	// Get instance from apiserver. Even though the instance doesn't have spec, we don't need to save
+	// this api call since otherwise an api read is required before writeInstance() anyway.
+	inst, err := r.getInstance(ctx, pnm, nm)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Report "Forbidden" state and early exist if the namespace is not allowed to self-serve
+	// namespaces but has bypassed the webhook and successfully created the hns instance.
+	// TODO refactor/split the EX map for 1) reconciler exclusion and 2) self-serve not allowed
+	//  purposes. See issue: https://github.com/kubernetes-sigs/multi-tenancy/issues/495
+	if EX[pnm] {
+		inst.Status.State = api.Forbidden
+		err := r.writeInstance(ctx, log, inst)
+		return ctrl.Result{}, err
+	}
+
+	// Get the self-serve namespace's hierarchyConfig and namespace instances.
+	hcInst, nsInst, err := r.hcr.GetInstances(ctx, log, nm)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	r.syncWithForest(log, inst, nsInst, hcInst)
+
+	// TODO report the "RequiredChildConflict" in the HNS reconciliation. See issue:
+	//  https://github.com/kubernetes-sigs/multi-tenancy/issues/490
+
+	err = r.writeInstance(ctx, log, inst)
+	return ctrl.Result{}, err
+}
+
+// HierarchicalNamespace (HNS) is synced with the in-memory forest, the HierarchyConfig and
+// the namespace instances to update its HNS state. This will be the only place the "Owner"
+// field in the forest is set. Therefore, the "Owner" field can be used in the forest to get
+// all the HNS objects of a namespace.
+func (r *HierarchicalNamespaceReconciler) syncWithForest(log logr.Logger, inst *api.HierarchicalNamespace, nsInst *corev1.Namespace, hcInst *api.HierarchyConfiguration) {
+	r.forest.Lock()
+	defer r.forest.Unlock()
+
+	// Names of the hierarchical namespace and the current namespace.
+	nm := inst.Name
+	pnm := inst.Namespace
+
+	// Get the namespace instance in memory.
+	ns := r.forest.Get(nm)
+
+	// If the real namespace instance doesn't exist yet, update forest and enqueue the namespace
```

Comment:
This function's getting a bit long. It might be nice to say something like:

```
switch {
case ns.Inst.Name == "":
  r.syncMissing(...)
case nsInst.Annotations[api.AnnotationOwner] != pnm:
  r.syncConflicting(...)
default:
  r.syncExisting(...)
}
```

That cleanly exposes the top-level logic of this function and allows each sub-case to be handled in its own function. sgty?